### PR TITLE
WIP: coco integration

### LINKED ISSUE #1069: Integrate COCO dataset
Perform benign evaluation on COCO dataset for DAPRICOT scenario. This is a practical use case for #1001 as the COCO dataset is available in the tfds catalog

### PR DESCRIPTION
**Pull request by lcadalzo**
Closes #1069.

Beyond integrating the COCO dataset, this PR also ensures that (1) the COCO dataset, (2) the APRICOT dataset, and (3) the COCO baseline model (which is used for APRICOT too) are all 0-indexed, meaning object labels start at 0 rather than 1. 

The tfds COCO dataset comes 0-indexed out-of-the-box, as does the [baseline model](https://github.com/lcadalzo/armory/blob/1069-coco-integration/armory/baseline_models/tf_graph/mscoco_frcnn.py#L22). The APRICOT dataset came 1-indexed, though, and so previously (i.e. currently in Armory 0.13.1), the baseline model was [modified](https://github.com/twosixlabs/armory/blob/v0.13.1/armory/baseline_models/tf_graph/mscoco_frcnn.py#L54-L61) to output 1-indexed classes to avoid a label mismatch. 

This PR undoes this modification and makes everything 0-indexed. For the APRICOT dataset, this means adding a label_preprocessing function to decrement labels by 1 so that they're 0-indexed and thus align with the baseline model (and COCO dataset).

Note: this baseline model is also used in the D-APRICOT scenario, although the D-APRICOT dataset doesn't contain labels for COCO objects so no update to the D-APRICOT dataset is needed

### CODE REVIEW

**Comment by lcadalzo** [comment]
Running the baseline model on the COCO dataset for 50 train samples, I noticed that the mAP is lower than we'd expect, below 0.1. Taking a closer look, the average precision (AP) for the lower-indexed classes is relatively high whereas AP is 0 for all classes > 9. We might expect mAP to be higher on lower indices since these are more common classes, but likely not to this extent.
```
AP per class: {0: 0.64, 1: 0.18, 2: 0.42, 3: 1.0, 4: 0.68, 5: 1.0, 6: 0.0, 7: 0.62, 8: 0.73, 9: 0.51, 10: 0.0, 14: 0.0, 15: 0.0, 16: 0.0, 18: 0.0, 19: 0.0, 20: 0.0, 21: 0.0, 22: 0.0, 23: 0.0, 24: 0.0, 25: 0.0, 26: 0.0, 27: 0.0, 29: 0.0, 30: 0.0, 31: 0.0, 32: 0.0, 33: 0.0, 34: 0.0, 35: 0.0, 36: 0.0, 37: 0.0, 38: 0.0, 39: 0.0, 40: 0.0, 41: 0.0, 42: 0.0, 43: 0.0, 45: 0.0, 46: 0.0, 48: 0.0, 50: 0.0, 51: 0.0, 52: 0.0, 53: 0.0, 55: 0.0, 56: 0.0, 57: 0.0, 58: 0.0, 59: 0.0, 60: 0.0, 61: 0.0, 62: 0.0, 63: 0.0, 64: 0.0, 66: 0.0, 67: 0.0, 69: 0.0, 72: 0.0, 73: 0.0, 74: 0.0, 75: 0.0, 76: 0.0, 78: 0.0, 79: 0.0, 80: 0.0, 81: 0.0, 83: 0.0, 84: 0.0, 85: 0.0}
mAP: 0.08140845070422535
```

I noticed that for some images, the model's predicted label differs from the ground-truth label in the dataset _but aligns with the ground-truth label per COCO label maps that can be found online_ (e.g. [here](https://github.com/tensorflow/models/blob/master/research/object_detection/data/mscoco_label_map.pbtxt) or [here](https://tech.amikelive.com/node-718/what-object-categories-labels-are-in-coco-dataset/)). To be precise, it aligns with the 0-indexed version of these label maps (i.e. subtract 1 from each label).

To give an example: a frisbee is labeled in the dataset as class 29. The model predicts class 33, and 33 is the 0-indexed label for frisbee in the two links above. 

Taking a closer look, the labels in the tfds dataset range from 0 - 79 with no gaps, but the baseline model appears to have been trained with the (linked above) label map that ranges from 0 - 89 with some indices unused. This misalignment between the model's label map and dataset label map is causing mAP to be very low. One solution could be, during label preprocessing, to map the 0-79 label space to 0-89 with the appropriate gaps.

**Comment by lcadalzo** [comment]
Code to reproduce this can be found below. Set the `PLOT_IMAGES_WITH_LABELS_AND_PREDS` variable to `True` to write output images with labels and predictions overlaid in cyan and green respectively. Otherwise it will simply iterate over `NUM_TEST_SAMPLES` images and print the AP per class as well as ground-truth label distribution

```
from collections import Counter
import itertools
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import numpy as np
from PIL import Image

from armory.baseline_models.tf_graph import mscoco_frcnn
from armory.data import datasets
from armory.utils.metrics import object_detection_AP_per_class


def draw_box(x, y, pred_labels):
    x_pil = Image.fromarray(np.uint8(x[0]  * 255.))
    fig, ax = plt.subplots()
    ax.imshow(x_pil)
    boxes = y[0]["boxes"]
    labels = y[0]["labels"]
    img_h, img_w = x.shape[1:3]
    for i in range(labels.size):
        box_i = boxes[i]
        xmin = box_i[1] * img_w
        ymin = box_i[0] * img_h
        box_h = (box_i[2] - box_i[0]) * img_h
        box_w = (box_i[3] - box_i[1]) * img_w
        rect = patches.Rectangle((xmin, ymin), box_w, box_h, linewidth=1, edgecolor='r', facecolor='none')
        ax.add_patch(rect)
        ax.annotate(str(labels[i]), (xmin, ymin), color="c")
    ax.annotate(f"pred labels: {str(pred_labels)}", (0, 0), color="g")
    plt.savefig("coco_img.jpg")
    breakpoint()

def main():
    NUM_TEST_SAMPLES = 250
    PLOT_IMAGES_WITH_LABELS_AND_PREDS = False

    ds = datasets.coco2017(split="train", shuffle_files=False, batch_size=1)
    model = mscoco_frcnn.get_art_model(model_kwargs={}, wrapper_kwargs={})
    y_list = []
    y_pred_list = []
    for _ in range(NUM_TEST_SAMPLES):
        x, y = ds.get_batch()
        y_pred = model.predict(x)
        pred_labels = y_pred[0]["labels"][np.where(y_pred[0]["scores"] > 0.5)[0]]
        y_list.extend(y)
        y_pred_list.extend(y_pred)
        if PLOT_IMAGES_WITH_LABELS_AND_PREDS:
            draw_box(x, y, pred_labels)

    ap_per_class = object_detection_AP_per_class(y_list, y_pred_list)
    print(f"AP per class: {ap_per_class}")
    print(f"mAP: {np.fromiter(ap_per_class.values(), dtype=float).mean()}")
    instances_per_class = Counter(list(itertools.chain(*[list(i["labels"]) for i in y_list])))
    print(f"number of ground-truth objects per class: {instances_per_class}")

if __name__ == "__main__":
    main()
```

**Comment by lcadalzo** [comment]
The addition of COCO tests are causing the CI to fail for what appears to be an issue with the host's resources

**Comment by lcadalzo** [comment]
Confirming that the TFDS COCO dataset labels range from 0-79 with no gaps:

```
>>> instances_per_class
Counter({0: 262465, 2: 43867, 56: 38491, 73: 24715, 39: 24342, 41: 20650, 60: 15714, 45: 14358, 9: 12884, 26: 12354, 25: 11431, 14: 10806, 8: 10759, 7: 9973, 13: 9838, 18: 9509, 46: 9458, 33: 9076, 3: 8725, 24: 8720, 58: 8652, 19: 8147, 40: 7913, 51: 7852, 43: 7770, 50: 7308, 54: 7179, 1: 7113, 30: 6646, 75: 6613, 17: 6587, 27: 6496, 67: 6434, 49: 6399, 55: 6353, 32: 6347, 74: 6334, 28: 6192, 44: 6165, 37: 6126, 5: 6069, 47: 5851, 53: 5821, 62: 5805, 57: 5779, 65: 5703, 71: 5610, 36: 5543, 20: 5513, 16: 5508, 42: 5479, 22: 5303, 4: 5135, 23: 5131, 63: 4970, 38: 4812, 77: 4793, 15: 4768, 6: 4571, 48: 4373, 59: 4192, 61: 4157, 35: 3747, 69: 3334, 34: 3276, 52: 2918, 66: 2855, 31: 2685, 29: 2682, 72: 2637, 64: 2262, 11: 1983, 79: 1954, 10: 1865, 68: 1673, 76: 1481, 21: 1294, 12: 1285, 70: 225, 78: 198})
>>> instances_per_class.keys()
dict_keys([14, 57, 0, 27, 65, 31, 4, 7, 34, 32, 35, 39, 71, 79, 67, 2, 17, 30, 23, 25, 72, 45, 60, 69, 38, 59, 73, 56, 41, 61, 29, 37, 6, 21, 33, 36, 63, 1, 5, 8, 50, 16, 12, 26, 22, 24, 13, 11, 43, 19, 9, 42, 55, 15, 66, 68, 76, 51, 3, 20, 28, 47, 49, 44, 48, 62, 74, 18, 10, 54, 58, 64, 52, 53, 75, 77, 40, 46, 70, 78])
>>> max(instances_per_class.keys())
79
>>> len(instances_per_class)
80

```

**Comment by lcadalzo** [comment]
After mapping labels to 0-89, the results look much more in line with what we'd expect:
```
AP_per_class: {0: 0.74, 1: 0.0, 2: 0.79, 5: 0.95, 7: 1.0, 8: 0.0, 9: 1.0, 10: 1.0, 12: 0.64, 14: 0.27, 15: 0.0, 16: 1.0, 18: 1.0, 20: 1.0, 21: 0.61, 22: 1.0, 24: 0.64, 26: 1.0, 27: 0.91, 30: 0.14, 32: 0.0, 34: 0.27, 35: 0.55, 38: 1.0, 39: 0.64, 40: 1.0, 42: 1.0, 43: 0.39, 45: 0.0, 46: 0.63, 47: 0.64, 48: 0.09, 49: 0.0, 50: 0.72, 52: 1.0, 53: 0.0, 55: 0.68, 56: 0.39, 57: 0.24, 58: 1.0, 61: 0.35, 62: 0.73, 63: 0.7, 64: 0.92, 66: 0.35, 69: 1.0, 71: 1.0, 72: 1.0, 73: 1.0, 74: 0.55, 75: 0.67, 76: 0.12, 77: 0.0, 78: 0.91, 79: 0.0, 80: 0.0, 81: 0.0, 83: 0.44, 84: 1.0, 85: 0.45, 87: 1.0}
mAP: 59.21%
```

**Comment by lcadalzo** [comment]
Closing due to #1097
